File: app/src/main/java/com/pixplicity/cryptogram/utils/KeyboardUtils.java
package com.pixplicity.cryptogram.utils;

import android.app.Activity;
import android.content.Context;
import android.support.annotation.DrawableRes;
import android.support.annotation.NonNull;
import android.view.KeyEvent;
import android.view.View;

public class KeyboardUtils {

    private enum KeyValue {
        BACKSPACE(com.pixplicity.cryptogram.R.drawable.ic_keyboard_backspace, KeyEvent.KEYCODE_DEL),
        TAB(com.pixplicity.cryptogram.R.drawable.ic_keyboard_tab, KeyEvent.KEYCODE_ENTER),
        A("A", KeyEvent.KEYCODE_A),
        B("B", KeyEvent.KEYCODE_B),
        C("C", KeyEvent.KEYCODE_C),
        D("D", KeyEvent.KEYCODE_D),
        E("E", KeyEvent.KEYCODE_E),
        F("F", KeyEvent.KEYCODE_F),
        G("G", KeyEvent.KEYCODE_G),
        H("H", KeyEvent.KEYCODE_H),
        I("I", KeyEvent.KEYCODE_I),
        J("J", KeyEvent.KEYCODE_J),
        K("K", KeyEvent.KEYCODE_K),
        L("L", KeyEvent.KEYCODE_L),
        M("M", KeyEvent.KEYCODE_M),
        N("N", KeyEvent.KEYCODE_N),
        O("O", KeyEvent.KEYCODE_O),
        P("P", KeyEvent.KEYCODE_P),
        Q("Q", KeyEvent.KEYCODE_Q),
        R("R", KeyEvent.KEYCODE_R),
        S("S", KeyEvent.KEYCODE_S),
        T("T", KeyEvent.KEYCODE_T),
        U("U", KeyEvent.KEYCODE_U),
        V("V", KeyEvent.KEYCODE_V),
        W("W", KeyEvent.KEYCODE_W),
        X("X", KeyEvent.KEYCODE_X),
        Y("Y", KeyEvent.KEYCODE_Y),
        Z("Z", KeyEvent.KEYCODE_Z),
        SPACE(" ", KeyEvent.KEYCODE_SPACE),;

        public final String text;
        public final int drawableRes;
        public final int keycode;

        KeyValue(String text, int keycode) {
            this.text = text;
            this.drawableRes = 0;
            this.keycode = keycode;
        }

        KeyValue(@DrawableRes int drawableRes, int keycode) {
            this.text = null;
            this.drawableRes = drawableRes;
            this.keycode = keycode;
        }

        @NonNull
        private static KeyValue fromKeycode(int keyIndex) {
            return KeyValue.values()[keyIndex];
        }
    }

    public static void dispatch(final Contract contract) {
        final View focusedView = ((Activity) contract.getContext()).getCurrentFocus();
        if (focusedView != null) {
            int keycode = getKeycode(contract);
            focusedView.dispatchKeyEvent(new KeyEvent(0, 0, KeyEvent.ACTION_DOWN,
                    keycode, 0));
            focusedView.dispatchKeyEvent(new KeyEvent(0, 0, KeyEvent.ACTION_UP,
                    keycode, 0));
        }
    }

    private static int getKeycode(Contract contract) {
        return KeyValue.fromKeycode(contract.getKeyIndex()).keycode;
    }

    public static String getKeyText(Contract contract) {
        return KeyValue.fromKeycode(contract.getKeyIndex()).text;
    }

    @DrawableRes
    public static int getKeyIcon(Contract contract) {
        return KeyValue.fromKeycode(contract.getKeyIndex()).drawableRes;
    }

    public interface Contract {

        Context getContext();

        void setOnClickListener(View.OnClickListener onClickListener);

        int getKeyIndex();

    }

}

File: app/src/main/java/com/pixplicity/cryptogram/utils/KeyboardUtils.kt
package com.pixplicity.cryptogram.utils

import android.support.annotation.DrawableRes
import android.view.KeyEvent
import com.pixplicity.cryptogram.events.PuzzleEvent

object KeyboardUtils {

    interface Contract {
        companion object {
            const val ALPHA_GREYED = (256 * 0.4f).toInt()
        }
        val keyIndex: Int
    }

    private enum class KeyValue(
            @DrawableRes val drawableRes: Int,
            val text: String?,
            val keycode: Int) {

        BACKSPACE(com.pixplicity.cryptogram.R.drawable.ic_keyboard_backspace, null, KeyEvent.KEYCODE_DEL),
        TAB(com.pixplicity.cryptogram.R.drawable.ic_keyboard_tab, null, KeyEvent.KEYCODE_ENTER),
        A(0, "A", KeyEvent.KEYCODE_A),
        B(0, "B", KeyEvent.KEYCODE_B),
        C(0, "C", KeyEvent.KEYCODE_C),
        D(0, "D", KeyEvent.KEYCODE_D),
        E(0, "E", KeyEvent.KEYCODE_E),
        F(0, "F", KeyEvent.KEYCODE_F),
        G(0, "G", KeyEvent.KEYCODE_G),
        H(0, "H", KeyEvent.KEYCODE_H),
        I(0, "I", KeyEvent.KEYCODE_I),
        J(0, "J", KeyEvent.KEYCODE_J),
        K(0, "K", KeyEvent.KEYCODE_K),
        L(0, "L", KeyEvent.KEYCODE_L),
        M(0, "M", KeyEvent.KEYCODE_M),
        N(0, "N", KeyEvent.KEYCODE_N),
        O(0, "O", KeyEvent.KEYCODE_O),
        P(0, "P", KeyEvent.KEYCODE_P),
        Q(0, "Q", KeyEvent.KEYCODE_Q),
        R(0, "R", KeyEvent.KEYCODE_R),
        S(0, "S", KeyEvent.KEYCODE_S),
        T(0, "T", KeyEvent.KEYCODE_T),
        U(0, "U", KeyEvent.KEYCODE_U),
        V(0, "V", KeyEvent.KEYCODE_V),
        W(0, "W", KeyEvent.KEYCODE_W),
        X(0, "X", KeyEvent.KEYCODE_X),
        Y(0, "Y", KeyEvent.KEYCODE_Y),
        Z(0, "Z", KeyEvent.KEYCODE_Z),
        SPACE(0, " ", KeyEvent.KEYCODE_SPACE)
    }

    private fun fromKeycode(keyIndex: Int): KeyValue {
        return KeyValue.values()[keyIndex]
    }

    fun dispatch(contract: Contract) {
        EventProvider.postEvent(
                PuzzleEvent.KeyboardInputEvent(getKeycode(contract)))
    }

    private fun getKeycode(contract: Contract): Int {
        return fromKeycode(contract.keyIndex).keycode
    }

    fun getKeyText(contract: Contract): String? {
        return fromKeycode(contract.keyIndex).text
    }

    @DrawableRes
    fun getKeyIcon(contract: Contract): Int {
        return fromKeycode(contract.keyIndex).drawableRes
    }

}
